diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt

```diff
@@ -56,15 +56,30 @@ class DataClassesRule(configRules: List<RulesConfig>) : DiktatRule("data-classes
     @Suppress("UnsafeCallOnNullableType", "FUNCTION_BOOLEAN_PREFIX", "ComplexMethod")
     private fun ASTNode.canBeDataClass(): Boolean {
         val isNotPropertyInClassBody = findChildByType(CLASS_BODY)?.let { (it.psi as KtClassBody).properties.isEmpty() } ?: true
+        val constructorParametersNames = mutableListOf<String>()
         val hasPropertyInConstructor = findChildByType(PRIMARY_CONSTRUCTOR)
             ?.let { constructor ->
                 (constructor.psi as KtPrimaryConstructor)
                     .valueParameters
+                    .onEach {
+                        if (!it.hasValOrVar())
+                            constructorParametersNames.add(it.name!!)
+                    }
                     .run { isNotEmpty() && all { it.hasValOrVar() } }
             } ?: false
         if (isNotPropertyInClassBody && !hasPropertyInConstructor) {
             return false
         }
+        // if parameter of the primary constructor is used in init block then it is hard to refactor this class to data class
+        if (constructorParametersNames.isNotEmpty()) {
+            val initBlocks = findChildByType(CLASS_BODY)?.getAllChildrenWithType(CLASS_INITIALIZER)
+            initBlocks?.forEach { init ->
+                val refExpressions = init.findAllNodesWithSpecificType(REFERENCE_EXPRESSION)
+                if (refExpressions.any { it.text in constructorParametersNames }) {
```

Comment:
Please add a test that would have passed when there was a mistake with `!in` and shouldn't actually pass.

Reply:
Added